nemo/collections/asr/models/hybrid_asr_tts_models.py

```diff
@@ -0,0 +1,459 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import itertools
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import torch
+from omegaconf import DictConfig, OmegaConf
+from pytorch_lightning import Trainer
+from torch.nn.utils.rnn import pad_sequence
+
+from nemo.collections.asr.data.audio_to_text_dali import DALIOutputs
+from nemo.collections.asr.data.text_to_text import (
+    TextOrAudioToTextBatch,
+    TextToTextBatch,
+    TextToTextDataset,
+    TextToTextIterableDataset,
+)
+from nemo.collections.asr.models import ASRModel, EncDecCTCModelBPE, EncDecRNNTBPEModel
+from nemo.collections.asr.modules.conformer_encoder import ConformerEncoder
+from nemo.collections.asr.parts.preprocessing.features import clean_spectrogram_batch, normalize_batch
+from nemo.collections.asr.parts.submodules.batchnorm import replace_bn_with_fused_bn_all
+from nemo.collections.common.data import ConcatDataset, ConcatMapDataset
+from nemo.collections.tts.models import FastPitchModel
+from nemo.core.classes import Dataset, ModelPT
+from nemo.core.classes.common import PretrainedModelInfo
+from nemo.utils import logging
+from nemo.utils.enum import PrettyStrEnum
+
+
+def _fuse_bn_in_conformer(asr_model: ASRModel):
+    """
+    Replace BatchNorm with Fused BatchNorm in Conformer and fixes model config inplace
+    Expected `encoder` model to exist and be of type ConformerEncoder
+    """
+    logging.info("Replacing BatchNorm with Fused BatchNorm")
+    if not hasattr(asr_model, "encoder"):
+        raise NotImplementedError("No encoder found in ASR Model, replacement not supported")
+    if not isinstance(asr_model.encoder, ConformerEncoder):
+        raise NotImplementedError(f"Unsupported encoder type: {type(asr_model.encoder)}")
+    replace_bn_with_fused_bn_all(asr_model.encoder)
+    if "conv_norm_type" not in asr_model.cfg.encoder:
+        # old CTC models from NGC don't have such param
+        logging.warning("conv_norm_type not in encoder config, adding parameter")
+        OmegaConf.set_struct(asr_model.cfg, False)
+        asr_model.cfg.encoder.conv_norm_type = "fused_batch_norm"
+        OmegaConf.set_struct(asr_model.cfg, True)
+    else:
+        asr_model.cfg.encoder.conv_norm_type = "fused_batch_norm"
+
+
+class ASRWithTTSModel(ASRModel):
+    """
+    Hybrid ASR-TTS model: a transparent wrapper for ASR model
+    with frozen text-to-spectrogram pretrained model, which allows to use text-only data for training/finetuning
+    Text-only data can be mixed with audio-text pairs
+    """
+
+    asr_model: Union[EncDecRNNTBPEModel, EncDecCTCModelBPE]
+    tts_model: FastPitchModel
+
+    class ASRModelTypes(PrettyStrEnum):
```

Comment:
Could you please explain why do we need this juggling?

Reply:
Using string literals (e.g. "rnnt_bpe") is not error-prone. This is the purpose why enums are used in different languages) 

Imagine we have a string from config provided by the user. 
We will need to check everywhere when condition on this value, that this value is valid. Enum provides a clear way to convert and check it only once: `model_type = cls.ASRModelTypes(asr_model_type)`, and in all other places we don't need to check, if it is correct.

Also, when I write something like
```python
if model_type == "rmnt_bpe":
    ...
    # do something specific to rnnt-bpe model
```
if I have a typo in a value, the code still will work (but incorretly). If I compare a value with non-existing enum value – I will immediately get an error (which is significantly better, since we can't test all the code logic).